### PR TITLE
【FlexCheckpoint】Add ShardedTensor, Upgrade DCP, and Introduce AOA

### PR DESCRIPTION
**Pull request by xingmingyyj**
### PR Category

User Experience
### PR Types

New features

### Description

添加FlexCheckpoint主要模块，包括
- ShardedTensor，为Tensor增加了切分信息描述。
- AOAEngine，解析AOA标记，提供目标Tensor切片和源Tensor切片的映射关系。
- 对原有的dist checkpoint接口做升级，可以直接对ShardedTensor做save，load。
pcard-73263

### CODE REVIEW

File: python/paddle/distributed/fleet/layers/mpu/mp_layers.py [hunk 1]
```diff
@@ -528,6 +540,15 @@ def _overlap_linear():
             output = output_parallel
         return output
 
+    def sharded_state_dict(
+        self,
+        structured_name_prefix: str = "",
+    ):
+        state_dict = self.state_dict(structured_name_prefix="")
+        return build_sharded_state_dict(
+            state_dict, {"weight": 1, "bias": 0}, structured_name_prefix
```
Comment (From00) [inline python/paddle/distributed/fleet/layers/mpu/mp_layers.py:547]: bias为None的情况也支持切分吗

File: python/paddle/distributed/flex_checkpoint/dcp/load_state_dict.py
Comment (AndSonder) [inline python/paddle/distributed/flex_checkpoint/dcp/load_state_dict.py:1]: 这个文件路径的修改会引起 paddleformers 等依赖库 import 报错，是否先兼容性修改，保留 checkpoint 文件夹下的 `__init__.py`, 待依赖库都修改后再删除这个路径
image

File: python/paddle/distributed/flex_checkpoint/dcp/reshard.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        keys = sorted(grouped.keys())
+        pos = global_ranges[0][0]
+        for start, end in keys:
+            if start != pos:
+                return False
+            if end <= start:
+                return False
+            pos = end
+        if pos != global_ranges[0][1]:
+            return False
```
Comment (From00) [inline python/paddle/distributed/flex_checkpoint/dcp/reshard.py:64]: 这几行和L41-L48基本是一样的，但为何一个判断`start != pos`，一个是`start > pos`。可否统一对`ndim=2`的代码做下复用。

File: python/paddle/distributed/flex_checkpoint/dcp/reshard.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        local_dst_state_dict_shard_info,
+        group=process_group,
+    )
+
+    dst_state_dict_shard_info = {}
+    for rank_shard_info in global_dst_state_dict_shard_info:
+        for key, tensor_shard_info in rank_shard_info.items():
+            if key not in dst_state_dict_shard_info:
+                dst_state_dict_shard_info[key] = []
+            dst_state_dict_shard_info[key].append(tensor_shard_info)
```
Comment (From00) [inline python/paddle/distributed/flex_checkpoint/dcp/reshard.py:247]: 这段代码和src_state_dict的也可以复用

File: python/paddle/distributed/flex_checkpoint/dcp/utils.py [hunk 4]
```diff
@@ -150,3 +154,60 @@ def check_unique_id(unique_id, process_group):
...
+    end_idx = unravel_index(flat_end - 1, shape)
+    min_slices = []
+    for axis in range(len(shape)):
+        if axis == 0:
+            s = start_idx[axis]
+            e = end_idx[axis] + 1
+        else:
+            if start_idx[axis - 1] == end_idx[axis - 1]:
+                s = min(start_idx[axis], end_idx[axis])
+                e = max(start_idx[axis], end_idx[axis]) + 1
```
Comment (From00) [inline python/paddle/distributed/flex_checkpoint/dcp/utils.py:185]: 什么情况下会出现`start_idx [axis]> end_idx[axis]`？另外，`start_idx[axis - 1] == end_idx[axis - 1]`的条件，应该只允许在`axis==1`的情况出现？所以这里unflatten的过程是不是可以简单抽象成：
```
min_slices.append((start_idx[0], end_idx[0]+1))
min_slices.append((0, shape[axis])) for axis in range(1, len(shape))
```

File: python/paddle/distributed/flex_checkpoint/dcp/save_state_dict.py [hunk 5]
```diff
@@ -163,8 +170,140 @@ def save_state_dict(
...
+            need_reshard = {}
+            for key, shard in flattened.items():
+                local_shape = shard.local_shape
+                flat_range = shard.flattened_range
+                flat_start, flat_end = flat_range.start, flat_range.stop
+                slices, start_idx, end_idx = minimal_nd_slice(
+                    local_shape, flat_start, flat_end
+                )
+                min_shape = tuple(e - s for s, e in slices)
+                min_offset = tuple(
```
Comment (From00) [inline python/paddle/distributed/flex_checkpoint/dcp/save_state_dict.py:194]: flattened_tensor的global_offset，不是在unflattened状态下标记的吗，为何还需要构造一个单独的min_offset？另外这里是不是只对于axis==0，min_offset和global_offset才会有不同？

File: python/paddle/distributed/flex_checkpoint/dcp/save_state_dict.py [hunk 6]
```diff
@@ -163,8 +170,140 @@ def save_state_dict(
...
+                        min_shape,
+                        min_offset,
+                        shard,
+                    )
+                    need_reshard[temp_key] = ShardedTensor(
+                        key=temp_key,
+                        local_tensor=tmp_tensor,
+                        local_shape=(numel,),
+                        global_shape=(math.prod(local_shape),),
+                        global_offset=(
```
Comment (From00) [inline python/paddle/distributed/flex_checkpoint/dcp/save_state_dict.py:225]: 这里的global_offset是一个标量？

File: python/paddle/distributed/flex_checkpoint/dcp/save_state_dict.py [hunk 7]
```diff
@@ -163,8 +170,140 @@ def save_state_dict(
...
+            sharded_state_dict = state_dict
+            flattened, unflattened = {}, {}
+            for key, shard in sharded_state_dict.items():
+                if getattr(shard, "is_flattened", False):
+                    flattened[key] = shard
+                else:
+                    unflattened[key] = shard
+            reshaped_shards = {}
+            need_reshard = {}
+            for key, shard in flattened.items():
```
Comment (From00) [inline python/paddle/distributed/flex_checkpoint/dcp/save_state_dict.py:186]: flattened的处理逻辑比较难以理解，建议对一些关键步骤可以添加必要的注释。

File: python/paddle/distributed/flex_checkpoint/aoa/parser.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+
+class Statement:
+    def __init__(self, left_vars, right_vars, attrs):
+        self.left_vars = left_vars  # List[Var]
+        self.right_vars = right_vars  # List[Var]
+        self.attrs = attrs  # List[Attribute]
+
+    def __repr__(self):
+        return f"Statement({self.left_vars} <- {self.right_vars}, attrs={self.attrs})"
```
Comment (From00) [inline python/paddle/distributed/flex_checkpoint/aoa/parser.py]:
```suggestion
        return f"Statement({self.left_vars} -> {self.right_vars}, attrs={self.attrs})"
```

File: python/paddle/distributed/flex_checkpoint/aoa/parser.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .lexer import Token, TokenType
+
+
+class Statement:
+    def __init__(self, left_vars, right_vars, attrs):
```
Comment (From00) [inline python/paddle/distributed/flex_checkpoint/aoa/parser.py:19]: 这里可以加一下`len(left_vars) > 1 && len(right_vars) > 1`的拦截

File: python/paddle/distributed/flex_checkpoint/aoa/parser.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    def __repr__(self):
+        return f"{self.key}={self.value!r}"
+
+
+class Parser:
+    """
+    AOA Grammar
+    PROGRAM   ::= { STATEMENT }
+
+    STATEMENT ::= VAR_LIST '->' VAR ',' ATTR_LIST       // meige
```
Comment (From00) [inline python/paddle/distributed/flex_checkpoint/aoa/parser.py:50]:
```suggestion
    STATEMENT ::= VAR_LIST '->' VAR ',' ATTR_LIST       // merge
```

File: python/paddle/distributed/flex_checkpoint/aoa/lexer.py [hunk 11]
```diff
@@ -0,0 +1,413 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+import re
+from enum import Enum, auto
+
+
+def macro(name, priority):
```
Comment (From00) [inline python/paddle/distributed/flex_checkpoint/aoa/lexer.py:20]: macro后续建议可以单独拆一个文件

**Review by From00 — DISMISSED** [review]
LGTM，一些建议可以在下个PR修改。

File: python/paddle/distributed/fleet/layers/mpu/mp_layers.py (hunk 1 above)
Comment (xingmingyyj) [reply python/paddle/distributed/fleet/layers/mpu/mp_layers.py:547]: 如果bias为None，这个weight应该不会出现在stata_dict中，build_sharded_state_dict的逻辑是遍历state_dict中的k,v构造ShardedTensor，所以即便用户做了标记但是bias为None也不会有问题

File: python/paddle/distributed/flex_checkpoint/dcp/reshard.py (hunk 2 above)
Comment (xingmingyyj) [reply python/paddle/distributed/flex_checkpoint/dcp/reshard.py:64]: 这里是代码问题，下一个PR统一修改

File: python/paddle/distributed/flex_checkpoint/dcp/reshard.py (hunk 3 above)
Comment (xingmingyyj) [reply python/paddle/distributed/flex_checkpoint/dcp/reshard.py:247]: 好的，下个PR统一修复

File: python/paddle/distributed/flex_checkpoint/dcp/utils.py (hunk 4 above)
Comment (xingmingyyj) [reply python/paddle/distributed/flex_checkpoint/dcp/utils.py:185]: 这里是适配了一下高维tensor的情况。当len(shape)==2时：原代码就退化为
```python
min_slices.append((start_idx[0], end_idx[0]+1))
min_slices.append((0, shape[axis])) for axis in range(1, len(shape))
```

File: python/paddle/distributed/flex_checkpoint/dcp/save_state_dict.py (hunk 5 above)
Comment (xingmingyyj) [reply python/paddle/distributed/flex_checkpoint/dcp/save_state_dict.py:194]: flattened_tensor的global_offset，是在unflattened状态下做的标记。造一个单独的min_offset的原因是为了将flattened后的tensor扩展成dcp可以描述的切片。在大多数场景下只对于axis==0，min_offset和global_offset才会有不同。但是，如果这个flatten的切片本身就是在原local_tensor的一行切出来的，此时global_offset和min_offset在其他axis上也会不同

File: python/paddle/distributed/flex_checkpoint/dcp/save_state_dict.py (hunk 7 above)
Comment (xingmingyyj) [reply python/paddle/distributed/flex_checkpoint/dcp/save_state_dict.py:186]: 好的

File: python/paddle/distributed/flex_checkpoint/aoa/parser.py (hunk 8 above)
Comment (xingmingyyj) [reply python/paddle/distributed/flex_checkpoint/aoa/parser.py]: 好的

File: python/paddle/distributed/flex_checkpoint/aoa/lexer.py (hunk 11 above)
Comment (xingmingyyj) [reply python/paddle/distributed/flex_checkpoint/aoa/lexer.py:20]: 好的，下个PR拆出

File: python/paddle/distributed/flex_checkpoint/aoa/parser.py (hunk 9 above)
Comment (xingmingyyj) [reply python/paddle/distributed/flex_checkpoint/aoa/parser.py:19]: 好的

File: python/paddle/distributed/flex_checkpoint/aoa/parser.py (hunk 10 above)
Comment (xingmingyyj) [reply python/paddle/distributed/flex_checkpoint/aoa/parser.py:50]: 好的

File: python/paddle/distributed/fleet/layers/mpu/mp_layers.py [hunk 12]
```diff
@@ -740,6 +761,15 @@ def forward(self, x):
 
         return output
 
+    def sharded_state_dict(
+        self,
+        structured_name_prefix: str = "",
+    ):
+        state_dict = self.state_dict(structured_name_prefix="")
+        return build_sharded_state_dict(
+            state_dict, {"weight": 0, "bias": 0}, structured_name_prefix
```
Comment (AndSonder) [inline python/paddle/distributed/fleet/layers/mpu/mp_layers.py]: RowParallel 应该不切 bias 吧

Comment (xingmingyyj) [reply python/paddle/distributed/fleet/layers/mpu/mp_layers.py]: 确实不切，下个PR一起修改

Comment (AndSonder) [reply python/paddle/distributed/fleet/layers/mpu/mp_layers.py]: > RowParallel

RowSeqParallel 也是同理需要修改 ～

Comment (xingmingyyj) [reply python/paddle/distributed/fleet/layers/mpu/mp_layers.py]: 好的好的

**Review by From00 — DISMISSED** [review]
LGTM

**Review by risemeup1 — COMMENTED** [review]
LGTM for setup.py

**Review by risemeup1 — DISMISSED** [review]

**Review by sunzhongkai588 — DISMISSED** [review]
LGTM， 
我看 ShardedTensor API 都暴露了，中文文档也补齐一下

**Review by zyfncg — DISMISSED** [review]
LGTM for setup.py.in

**Review by From00 — DISMISSED** [review]

**Review by zyfncg — DISMISSED** [review]

**Review by risemeup1 — DISMISSED** [review]

**Review by raindrops2sea — DISMISSED** [review]

**Review by From00 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/distributed/__init__.py [hunk 13]
```diff
@@ -229,4 +235,8 @@
     "ContextParallel",
     "PrepareContextParallel",
     "create_nccl_config",
+    "ShardedWeight",
+    "ShardedStateDict",
```
Comment (sunzhongkai588) [inline python/paddle/distributed/__init__.py:239]: `ShardedStateDict` 似乎没有文档？

**Review by sunzhongkai588 — APPROVED** [review]
LGTM，文档问题后续补充
